### PR TITLE
MOD 6431: [GeoShape] memory allocation patterns

### PR DESCRIPTION
**Pull request by ephraimfeldblum**
**Describe the changes in the pull request**

A clear and concise description of what the PR is solving, including:
Currently, GeoShape contains overloaded raw `new`/`delete`. Additionally, querying uses unnecessary excess memory to perform an eager transform on the returned docs.
Solved by adding more idiomatic memory management using `RediSearch::Allocator`. Adding `boost::allocate_unique` to fix a `std` oversight, enabling creation of `std::unique_ptr` using an appropriate allocator. Lazily transform query results, only storing them when ready to pass across an FFI boundary.

**Which issues this PR fixes**
1. [MOD-6431](https://redislabs.atlassian.net/browse/MOD-6431)

**Main objects this PR modified**
1. src/geometry/*

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

[MOD-6293]: https://redislabs.atlassian.net/browse/MOD-6293?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[MOD-6431]: https://redislabs.atlassian.net/browse/MOD-6431?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by GuyAv46 — COMMENTED** [review]
The code looks good.
Can you explain in the description what caused the segfault, and what was fixed in the test? It looks like only cosmetic changes to me

**Comment by ephraimfeldblum** [comment]
> The code looks good. Can you explain in the description what caused the segfault, and what was fixed in the test? It looks like only cosmetic changes to me

Nope.
If you take a look at the links I left in MOD-6293, it appears that the segfault is a known Clang sanitizer bug. There may be some workarounds we can try, but for the most part, they have little if anything to do with the actual sanitization of the code being run.

**Review by oshadmi — APPROVED** [review]

File: src/geometry/rtree.cpp [hunk 1]
```diff
@@ -99,9 +99,12 @@ constexpr auto geometry_reporter = [](auto const& geom) -> std::size_t {
     const auto outer_size = geom.outer().get_allocator().report();
     // reduce allows for out-of-order execution of associative and commutative operation (std::plus)
     // transform to associative and commutative using a unary predicate (hole -> size_t)
-    return std::transform_reduce(std::execution::unseq, std::begin(inners), std::end(inners),
-                                 outer_size, std::plus{},
-                                 [](auto const& hole) { return hole.get_allocator().report(); });
+    return std::transform_reduce(
+#ifndef __APPLE__  // apple clang does not implement `std::execution` despite being a C++17 feature
```
Comment (oshadmi) [inline src/geometry/rtree.cpp:103]: `__APPLE__` is only defined in clang? With gcc will use `std::execution` ?

Comment (ephraimfeldblum) [reply src/geometry/rtree.cpp:103]: `__APPLE__` is defined in apple. do we use `gcc` to compile on apple?

**Review by oshadmi — APPROVED** [review]

**Review by oshadmi — APPROVED** [review]

**Pull request merged**